pkg/cgroup/linux.go

```diff
@@ -0,0 +1,213 @@
+	dockerPrefixName = "/docker/"
+
+	// Proc controller group path.
+	procCGroupPath = "/proc/%d/cgroup"
+)
+
+// CGEntries - represents all the entries in a process cgroup file
+// at /proc/<pid>/cgroup.
+type CGEntries map[string][]string
+
+// GetCGPathByName - returns the path matching the controller name.
+func (cg CGEntries) GetCGPathByName(cname string) (path string) {
+	return cg[cname][0]
+}
+
+// GetProcCGroup - reads and parses all the cgroup entries for a given process.
+func GetProcCGroup(pid int) (CGEntries, error) {
+	r, err := os.Open(fmt.Sprintf(procCGroupPath, pid))
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	return parseProcCGroup(r)
+}
+
+// ParseProcCGroup - cgroups are always in the following
+// format once enabled you need to know the pid of the
+// application you are looking for so that the the
+// following parsing logic only parses the file located
+// at /proc/<pid>/cgroup.
+//
+// CGROUP entries id, component and path are always in
+// the following format. ``ID:COMPONENT:PATH``
+//
+// Following code block parses this information and
+// returns a procCGroup which is a parsed list of all
+// the line by line entires from /proc/<pid>/cgroup.
+func parseProcCGroup(r io.Reader) (CGEntries, error) {
+	var proc = make(CGEntries)
+
+	// Start reading cgroup categories line by line
+	// and process them into procCGroup structure.
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		tokens := strings.SplitN(line, ":", 3)
+		if len(tokens) < 3 {
+			continue
+		}
+
+		_, name, path := tokens[0], tokens[1], tokens[2]
+		for _, token := range strings.Split(name, ",") {
+			name = strings.TrimPrefix(token, "name=")
+			proc[name] = []string{path}
+		}
+	}
+
+	// Return upon any error while reading the cgroup categories.
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return proc, nil
+}
+
+// Control group manager CLI defaults.
+const (
+	// Change this if at all in future control group manager
+	// changes their cli name.
+	cgmProgName = "cgm"
+
+	// Action param name used to fetch value for a particular
+	// controller.
+	cgmProgGetValue = "getvalue"
+)
+
+// Fetch value of the cgroup kernel param from the cgroup manager,
+// if cgroup manager is configured we should just rely on `cgm` cli
+// to fetch all the values for us.
+func getCGroupManagerKernValue(cname, path, kernParam string) (limit uint64, err error) {
+	// Default value for an unconfigured cgroup memory limit.
+	//
+	//    $ cat /sys/fs/cgroup/memory/memory.limit_in_bytes
+	//    9223372036854771712
+	//
+	limit = unlimitedMemoryLimit
+
+	cmd := exec.Command(cgmProgName, cgmProgGetValue, cname, path, kernParam)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	if err = cmd.Run(); err != nil {
+		return limit, err
+	}
+
+	var v uint64
+	v, err = strconv.ParseUint(strings.TrimSpace(out.String()), 10, 64)
+	if err != nil {
+		return limit, err
+	}
+
+	limit = v
+	return limit, nil
```

Comment:
Below code is enough, as no one supposes to use `limit` in case of error is set.

```go
limit, err = strconv.ParseUint(strings.TrimSpace(out.String()), 10, 64)
return limit, err
```

little line saver is
```go
    var out bytes.Buffer
    cmd := exec.Command(cgmProgName, cgmProgGetValue, cname, path, kernParam)
    cmd.Stdout = &out
    if err = cmd.Run(); err == nil {
        limit, err = strconv.ParseUint(strings.TrimSpace(out.String()), 10, 64)
    }

    return limit, err
```